### PR TITLE
detect eth1 nodes that are not fully synced

### LINKED ISSUE #3337: Issues running Teku with OpenEthereum
### Description
We've got a user reporting the following stack trace when running Teku with [OpenEthereum](https://openethereum.wiki/Getting-Synced/).  It appears that the eth1 node is missing historical blocks, which is causing teku to throw a NPE:
```
16:17:20.699 INFO  - Eth1 tracker successfully caught up to chain head
16:17:20.707 FATAL - Exiting due to fatal error in Eth1DepositManager
java.util.concurrent.CompletionException: tech.pegasys.teku.service.serviceutils.FatalServiceFailureException: java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:990) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniExceptionally.tryFire(CompletableFuture.java:970) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture.uniWhenCompleteStage(CompletableFuture.java:883) ~[?:?]
        at java.util.concurrent.CompletableFuture.whenComplete(CompletableFuture.java:2315) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:503) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.whenComplete(SafeFuture.java:28) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.propagateResult(SafeFuture.java:147) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$exceptionallyCompose$25(SafeFuture.java:347) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at tech.pegasys.teku.infrastructure.async.SafeFuture.lambda$propagateResult$3(SafeFuture.java:152) ~[teku-infrastructure-async-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniWhenComplete(CompletableFuture.java:859) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniWhenComplete.tryFire(CompletableFuture.java:837) ~[?:?]
        at java.util.concurrent.CompletableFuture.postComplete(CompletableFuture.java:506) ~[?:?]
        at java.util.concurrent.CompletableFuture.complete(CompletableFuture.java:2137) ~[?:?]
        at org.web3j.utils.Async.lambda$run$1(Async.java:38) ~[core-4.6.2.jar:?]
        at java.util.concurrent.CompletableFuture$AsyncRun.run(CompletableFuture.java:1800) [?:?]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130) [?:?]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:630) [?:?]
        at java.lang.Thread.run(Thread.java:832) [?:?]
Caused by: tech.pegasys.teku.service.serviceutils.FatalServiceFailureException: java.util.concurrent.CompletionException: java.lang.NullPointerException
        at tech.pegasys.teku.pow.Eth1DepositManager.lambda$start$3(Eth1DepositManager.java:74) ~[teku-pow-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture.uniExceptionally(CompletableFuture.java:986) ~[?:?]
        ... 30 more
Caused by: java.util.concurrent.CompletionException: java.lang.NullPointerException
        at java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314) ~[?:?]
        at java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319) ~[?:?]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1155) ~[?:?]
        ... 29 more
Caused by: java.lang.NullPointerException
        at tech.pegasys.teku.pow.MinimumGenesisTimeBlockFinder.compareBlockTimestampToMinGenesisTime(MinimumGenesisTimeBlockFinder.java:87) ~[teku-pow-20.11.1.jar:20.11.1]
        at tech.pegasys.teku.pow.MinimumGenesisTimeBlockFinder.lambda$binarySearchLoop$0(MinimumGenesisTimeBlockFinder.java:60) ~[teku-pow-20.11.1.jar:20.11.1]
        at java.util.concurrent.CompletableFuture$UniCompose.tryFire(CompletableFuture.java:1146) ~[?:?]
        ... 29 more
Teku is shutting down
16:17:20.727 INFO  - Stopped ServerConnector@52af750b{HTTP/1.1, (http/1.1)}{0.0.0.0:5051}
16:17:20.730 INFO  - Stopping Javalin ...
16:17:20.730 INFO  - Javalin has stopped
16:17:20.732 WARN  - Error shutting down connection manager
```

We should try to make teku's eth1 fetching logic more forgiving, if possible.

### LINKED ISSUE #3357: Try to detect eth1 nodes that are not fully synced
### Description
We should add some logic in our PoW / eth1 code that tries to detect whether or not the configured eth1 node is in sync.  Troubleshooting problems with eth1 nodes will be simpler if we can rule out problems related to being out-of-sync.  

Additionally, if we know the eth1 node is in sync, we can avoid making unnecessary eth1 requests.  For example,  [`Eth1DepositManager` starts up](https://github.com/ConsenSys/teku/blob/0e3cc8bfff559fe5b92ce21da39b6518d7b67f9b/pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java#L73-L76) and tries to locate the minimum possible eth1 block that could trigger genesis.  If the eth1 node's latest block is before the minimum possible genesis time, we start requesting blocks 1 by 1 until we find the min genesis block.  If the eth1 node is very far behind the actual eth1 block that triggers genesis, we could end up requesting an excessive number of blocks.  

Relates to #3337

### PR DESCRIPTION
**Pull request by tbenr**
## PR Description

Include `isSyncing` method in `Eth1Provider` interface and relative implementations.

When requesting eht1 head in  `tech.pegasys.teku.pow.Eth1DepositManager#getHead`, a new check is added to verify if eth1 node is still syncing. if yes introduces a 30s delay before trying again.

Associated tests have been updated. A new test case has been added to cover the retry code flow. 

new costant: Costants.ETH1_SYNCING_RETRY_TIMEOUT

## Fixed Issue(s)
fixes #3357
fixes #3337

### CODE REVIEW

File: pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java [hunk 1]
```diff
@@ -165,7 +165,8 @@ private void preGenesisSubscription(final BigInteger lastProcessedBlockNumber) {
   }
 
   private SafeFuture<EthBlock.Block> getHead() {
-    return getLatestEth1BlockNumber()
+    return waitForEth1ProviderToBeInSync()
+        .thenCompose((__) -> getLatestEth1BlockNumber())
```
Comment (ajsutton) [inline pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java]: nit: I think the brackets are unnecessary here:
```suggestion
        .thenCompose(__ -> getLatestEth1BlockNumber())
```

File: pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java [hunk 2]
```diff
@@ -183,6 +184,22 @@ private void preGenesisSubscription(final BigInteger lastProcessedBlockNumber) {
...
+        .thenCompose(
+            isSyncing -> {
+              if (isSyncing) {
+                LOG.info(
+                    "Eth1DepositManager failed to get the head of Eth1: still syncing. Retrying in {} seconds.",
+                    Constants.ETH1_SYNCING_RETRY_TIMEOUT.toSeconds());
+                return asyncRunner
+                    .getDelayedFuture(Constants.ETH1_SYNCING_RETRY_TIMEOUT)
+                    .thenCompose(__ -> waitForEth1ProviderToBeInSync());
+              } else return SafeFuture.COMPLETE;
```
Comment (ajsutton) [inline pow/src/main/java/tech/pegasys/teku/pow/Eth1DepositManager.java]: We can simplify this a bit with `asyncRunner.runAfterDelay`.  And we always use braces on code blocks.
```suggestion
                return asyncRunner.runAfterDelay(
                    this::waitForEth1ProviderToBeInSync, Constants.ETH1_SYNCING_RETRY_TIMEOUT);
              } else {
                return SafeFuture.COMPLETE;
              }
```

**Review by ajsutton — COMMENTED** [review]
Looks really good and works well.  Nice job.  I've just left a couple of little suggestions in terms of code style but this is a great first PR.

**Comment by tbenr** [comment]
> Looks really good and works well.  Nice job.  I've just left a couple of little suggestions in terms of code style but this is a great first PR.

Thanks @ajsutton ! Applying suggestions. I use to squash the commits in PRs. Do you require single-commit PRs here too?

**Comment by ajsutton** [comment]
> Thanks @ajsutton ! Applying suggestions. I use to squash the commits in PRs. Do you require single-commit PRs here too?

No its fine, we do a squash when we merge the PR.

**Review by ajsutton — APPROVED** [review]
LGTM. Thanks for this - great work.

**Pull request merged**